CHE-21296 edits (#7)
benoitf/che-docs: modules/end-user-guide/pages/adding-a-vscode-extension.adoc

> :navtitle: Adding a Visual Studio Code extension to a workspace

Author: :navtitle: Adding a Visual Studio Code plug-in to a workspace

Reviewer: FYI VS Code has extensions, not plug-ins https://code.visualstudio.com/docs/editor/extension-marketplace

:keywords: Visual Studio Code extension, user-guide

[id="adding-visual-studio-code-extension_{context}"]

> = Adding a Visual Studio Code plug-in to a workspace

Reviewer: ```suggestion
= Adding a Visual Studio Code extension to a workspace
```

> Devfiles in the devfile v2 format don't contain any IDE-specific plug-ins. Instead, the plug-ins are described in a specific meta-folder.

Author: Devfiles in the devfile v2 format don't contain any IDE-specific plug-ins or Visual Studio Code extensions. Instead, the plug-ins and extensions are described in a specific meta-folder.

[id="plug-in-registries"]
== {prod-short} plug-in registries overview

> {prod-short} has a default plug-in registry installed in every {prod-short} instance. The Che-Theia IDE searches and installs the extensions from that location. 

Author: {prod-short} has a default plug-in and extensions registry installed in every {prod-short} instance. The Che-Theia IDE searches and installs the plug-ins and extensions from that location.

> Check this {prod-short} https://github.com/eclipse-che/che-plugin-registry[registry project] for an overview of default  plug-ins and source code. An online instance that is refreshed after every commit to the main branch, is located https://eclipse-che.github.io/che-plugin-registry/main/v3/plugins/[here]. You can use a different plug-in registry for Che-Theia if you don't work in air gap mode: only the default registry is available there.

Reviewer: ```suggestion
Check this {prod-short} https://github.com/eclipse-che/che-plugin-registry[registry project] for an overview of default plug-ins and source code. An online instance that is refreshed after every commit to the main branch, is located https://eclipse-che.github.io/che-plugin-registry/main/v3/plugins/[here]. You can use a different plug-in registry for Che-Theia if you don't work in air gap mode: only the default registry is available there.
```

> Check this {prod-short} https://github.com/eclipse-che/che-plugin-registry[registry project] for an overview of default plug-ins and source code. An online instance that is refreshed after every commit to the main branch, is located https://eclipse-che.github.io/che-plugin-registry/main/v3/plugins/[here]. You can use a different plug-in registry for Che-Theia if you don't work in air gap mode: only the default registry is available there.

Author: Check this {prod-short} https://github.com/eclipse-che/che-plugin-registry[registry project] for an overview of default plug-ins, extensions, and source code. An online instance that is refreshed after every commit to the main branch, is located https://eclipse-che.github.io/che-plugin-registry/main/v3/plugins/[here]. You can use a different plug-in or extension registry for Che-Theia if you don't work in air gap mode: only the default registry is available there.

> The plug-in overview for Che-Code Visual Studio Code editor is located in the https://www.open-vsx.org/[OpenVSX instance]. Air gap mode is not yet supported for this editor.

Author: The plug-in and extension overview for Che-Code Visual Studio Code editor is located in the https://www.open-vsx.org/[OpenVSX instance]. Air gap mode is not yet supported for this editor.

[id="visual-studio-code-extensions-json"]
== Adding an extension to .vscode/extensions.json

> The easiest way to add a Visual Studio Code plug-in to a workspace is to add it to the `.vscode/extensions.json` file. The main advantage of this method is that it works with all supported {prod-short} IDEs.

Author: The easiest way to add a Visual Studio Code extension to a workspace is to add it to the `.vscode/extensions.json` file. The main advantage of this method is that it works with all supported {prod-short} IDEs.

> If you use the Che-Theia IDE, the plug-in is installed and configured automatically. If you use a different suported IDE with the Che-Code Visual Studio Code fork, a pop-up with a recommendation to install the plug-in shows when you start the workspace.

Reviewer: ```suggestion
If you use the Che-Theia IDE, the extension is installed and configured automatically. If you use a different supported IDE with the Che-Code Visual Studio Code fork, a pop-up with a recommendation to install the plug-in shows when you start the workspace.
```

> If you use the Che-Theia IDE, the extension is installed and configured automatically. If you use a different supported IDE with the Che-Code Visual Studio Code fork, a pop-up with a recommendation to install the plug-in shows when you start the workspace.

Author: If you use the Che-Theia IDE, the extension is installed and configured automatically. If you use a different supported IDE with the Che-Code Visual Studio Code fork, a pop-up with a recommendation to install the extension shows when you start the workspace.

.Prerequisites
. You have the `.vscode/extensions.json` file in the root of the GitHub repository.

.Procedure

> . Add the plug-in id to `.vscode/extensions.json` file. Use a `.` sign to separate the publisher and plug-in names:

Author: Add the extension id to `.vscode/extensions.json` file. Use a `.` sign to separate the publisher and extension names:

    ]
  }
----

*NOTE* 

> If the specified set of plug-in ids isn't available in the {prod-short} registry, the workspace starts without the plug-in.

Reviewer: ```suggestion
If the specified set of extension ids isn't available in the {prod-short} registry, the workspace starts without the plug-in.
```

> If the specified set of extension ids isn't available in the {prod-short} registry, the workspace starts without the plug-in.

Author: If the specified set of extension ids isn't available in the {prod-short} registry, the workspace starts without the extension.

[id="che-theia-plug-ins-YAML"]
== Adding plug-in parameters to .che/che-theia-plugins.yaml
You can add extra parameters to a plug-in by modifying the `.che/che-theia-plugins.yaml` file. These modifications include:
* Defining the plug-ins for workspace installation.
* Changing default memory limit.
* Overriding default preferences.

=== Defining the plug-ins for workspace installation
Define the plug-ins to be installed in the workspace.

.Prerequisites
. You have the `.che/che-theia-plugins.yaml` file in the root of the GitHub repository.

.Procedure

> . Add the id of the plug-in in the  `.che/che-theia-plugins.yaml` file. Use the `/` sign to separate the publisher and plug-in name:

Reviewer: ```suggestion
. Add the id of the extension in the `.che/che-theia-plugins.yaml` file. Use the `/` sign to separate the publisher and plug-in name:
```

> . Add the id of the plug-in in the  `.che/che-theia-plugins.yaml` file. Use the `/` sign to separate the publisher and plug-in name:

Reviewer 2: ```suggestion
. Add the `id` of the plug-in in the  `.che/che-theia-plugins.yaml` file. Use the `/` sign to separate the publisher and plug-in name:
```
@benoitf, let's use 'extension' in the context of VS Code and 'plug-in' in the context of Che. In this sentence particularly, it would seem strange to be adding an 'extension' into the `che-theia-plugins.yaml` file :)

Reviewer: ok 👍

+
*EXAMPLE*
[source,yaml,subs="+quotes"]
----
- id: redhat/java  
----

=== Changing default memory limit

Override container settings such as the memory limit.

.Prerequisites
. You have the `.che/che-theia-plugins.yaml` file in the root of the GitHub repository.

.Procedure

> . Add an `override` section in the `.che/che-theia-plugins.yaml` file under the `id` of the plug-in.

Reviewer: ```suggestion
. Add an `override` section in the `.che/che-theia-plugins.yaml` file under the `id` of the extension.
```

> . Specify the memory limit for the extension. In the following example, {prod-short} automatically installs the Red Hat Visual Studio Code Java plug-in in the {prod-short} workspace and increases the memory of the workspace by two gigabytes:

Reviewer: ```suggestion
. Specify the memory limit for the extension. In the following example, {prod-short} automatically installs the Red Hat Visual Studio Code Java extension in the {prod-short} workspace and increases the memory of the workspace by two gigabytes:
```

      sidecar:  
        memoryLimit: 2Gi
----

=== Overriding default preferences

> Override the default preferences of the Visual Studio Code plug-ins for the workspace.

Reviewer: ```suggestion
Override the default preferences of the Visual Studio Code extension for the workspace.
```

.Prerequisites
. You have the `.che/che-theia-plugins.yaml` file in the root of the GitHub repository.

.Procedure

> . Add an `override` section in the `.che/che-theia-plugins.yaml` file under the `id` of the plug-in. 

Reviewer: ```suggestion
. Add an `override` section in the `.che/che-theia-plugins.yaml` file under the `id` of the extension. 
```

. Specify the preferences in the `Preferences` section. In the following example, {prod-short} automatically installs Red Hat Visual Studio Code Java extension in the {prod-short} workspace and sets the `java.server.launchMode` preference to `LightWeight`.
+
*EXAMPLE*
[source,yaml,subs="+quotes"]
----
  - id: redhat/java  
    override:  
      preferences:
        java.server.launchMode: LightWeight
----

*NOTE* You can also define the preferences in the `.vscode/settings.json` file, either by changing the preferences in the UI of your IDE or by manually adding them to the `.vscode/settings.json` file:
[source,json,subs="+quotes"]
----
{
"my.preferences": "my-value"
}
----

[id="visual-studio-code-extensions-in-devfile"]

> == Defining Visual Studio Code plug-in attributes in the devfile

Reviewer: ```suggestion
== Defining Visual Studio Code extension attributes in the devfile
```

> In case it's not possible to add extra files in the GitHub repository, you can define some of the plug-in attributes by inlining them in the devfile. You can use this procedure with both `.vscode/extensions.json` and `.che/che-theia-plugins.yaml` file contents.

Author: In case it's not possible to add extra files in the GitHub repository, you can define some of the plug-in or extension attributes by inlining them in the devfile. You can use this procedure with both `.vscode/extensions.json` and `.che/che-theia-plugins.yaml` file contents.

=== Inlining .vscode/extensions.json file

> Use `.vscode/extensions.json` file contents to inline plug-in attributes in the devfile.

Reviewer: ```suggestion
Use `.vscode/extensions.json` file contents to inline extension attributes in the devfile.
```

.Procedure
. Add an `attributes` section in your `devfile.yaml` file.
. Add `.vscode/extensions.json` in the `atributes` section. Add a `|` sign after the colon separator.

> . Paste the contents of the `.vscode/extensions.json` file after the `|` sign. The following example uses Red Hat Visual Studio Code Java plug-in attributes:

Reviewer: ```suggestion
. Paste the contents of the `.vscode/extensions.json` file after the `|` sign. The following example uses Red Hat Visual Studio Code Java extension attributes:
```
